### PR TITLE
ref: Rollup of Rust changes

### PR DESCRIPTION
**Pull request by untitaker**
#sync-getsentry

### CODE REVIEW

File: src/sentry/constants.py [hunk 1]
```diff
@@ -167,7 +167,7 @@ def get_all_languages():
     'project', 'errors', 'event_id', 'message', 'checksum', 'culprit', 'fingerprint', 'level',
     'time_spent', 'logger', 'server_name', 'site', 'received', 'timestamp', 'extra', 'modules',
     'tags', 'platform', 'release', 'dist', 'environment', 'transaction', 'key_id', '_meta',
-    'applecrashreport', 'device', 'repos', 'query', 'type', 'hashes',
```
Comment (untitaker) [inline src/sentry/constants.py]: This is such that `repos` is discarded as unknown attribute.

File: src/sentry/event_manager.py [hunk 2]
```diff
@@ -687,7 +688,7 @@ def stringify(f):
         elif '_meta' in data:
             del data['_meta']
 
-        self._data = prune_empty_keys(data)
+        self._data = CanonicalKeyDict(prune_empty_keys(data))
```
Comment (untitaker) [inline src/sentry/event_manager.py]: I don't think this does anything but we traditionally return `CanonicalKeyDict` from `EventManager`

File: src/sentry/constants.py (hunk 1 above)
Comment (jan-auer) [reply src/sentry/constants.py]: Updated this part to remove legacy interfaces without an error message (see `CLIENT_IGNORED_ATTRS`).

File: src/sentry/interfaces/stacktrace.py [hunk 3]
```diff
@@ -221,6 +221,9 @@ def slim_frame_data(frames, frame_allowance=settings.SENTRY_MAX_STACKTRACE_FRAME
     Removes various excess metadata from middle frames which go beyond
     ``frame_allowance``.
     """
+    if not RUST_ENABLE_TRIMMING:
```
Comment (jan-auer) [inline src/sentry/interfaces/stacktrace.py]: We should remove this and diff again. This needs to be implemented in Rust.

File: src/sentry/interfaces/breadcrumbs.py [hunk 4]
```diff
@@ -83,7 +70,10 @@ def to_json(self):
     @classmethod
     def normalize_crumb(cls, crumb):
         ty = crumb.get('type') or 'default'
-        level = crumb.get('level') or 'info'
+        level = crumb.get('level')
+        if level not in LOG_LEVELS_MAP or level == 'log':
```
Comment (mitsuhiko) [inline src/sentry/interfaces/breadcrumbs.py]: Isn't `log` not in `LOG_LEVELS_MAP` anyways?

File: src/sentry/interfaces/stacktrace.py [hunk 5]
```diff
@@ -353,16 +353,20 @@ def to_python(cls, data, raw=False):
...
-            pre_context, post_context = None, None
+            pre_context = None
+
+        post_context = data.get('post_context', None)
+        if isinstance(post_context, list) and post_context:
+            post_context = [c or '' for c in post_context]
+        else:
+            post_context = None
+
+        if not context_line and (pre_context or post_context):
```
Comment (mitsuhiko) [inline src/sentry/interfaces/stacktrace.py]: Should this not be `pre_context is not None or post_context is not None`?

File: tests/sentry/event_manager/test_normalization.py [hunk 6]
```diff
@@ -211,7 +196,7 @@ def test_logger():
     manager.normalize()
     data = manager.get_data()
     assert data['logger'] == DEFAULT_LOGGER_NAME
-    assert not any(e.get('name') == 'logger' for e in data['errors'])
+    assert not any(e.get('name') == 'logger' for e in data.get('errors', []))
```
Comment (mitsuhiko) [inline tests/sentry/event_manager/test_normalization.py]: Should this not be `data.get('errors') or ()`?

**Review by mitsuhiko — CHANGES_REQUESTED** [review]

File: src/sentry/interfaces/breadcrumbs.py (hunk 4 above)
Comment (jan-auer) [reply src/sentry/interfaces/breadcrumbs.py]: No, it only contains the standard names: sample, debug, info, warning, error, fatal

File: src/sentry/interfaces/stacktrace.py (hunk 5 above)
Comment (untitaker) [reply src/sentry/interfaces/stacktrace.py]: any falsy pre_context/post_context is coerced to `None` before

File: src/sentry/interfaces/breadcrumbs.py (hunk 4 above)
Comment (mitsuhiko) [reply src/sentry/interfaces/breadcrumbs.py]: Then the `or` can go.

File: src/sentry/interfaces/stacktrace.py (hunk 5 above)
Comment (mitsuhiko) [reply src/sentry/interfaces/stacktrace.py]: Ah yep.

File: src/sentry/interfaces/breadcrumbs.py (hunk 4 above)
Comment (untitaker) [reply src/sentry/interfaces/breadcrumbs.py]: fixed

File: tests/sentry/event_manager/test_normalization.py (hunk 6 above)
Comment (untitaker) [reply tests/sentry/event_manager/test_normalization.py]: Sure, would be more robust but this is a testcase

**Review by mitsuhiko — APPROVED** [review]

**Pull request merged**